DSL2 release

yuukiiwa opened the pull request:
One final check before the release!!
- [x] tested and slightly modified the original dsl2 implementation to [produce consistent results](https://github.com/nf-core/nanoseq/files/7441305/compare_dsl1_dsl2.tar.gz) with the dsl1 pipeline
- [x] solved bug on [alignment being skipped when providing `--skip_demultiplexing`](https://github.com/nf-core/nanoseq/issues/77#issuecomment-952106691)
    - Here is [the log of running test_bc_nodx.log](https://github.com/nf-core/nanoseq/files/7441282/bc_nodx_alignment_ran_through.log)
- [x] replaced current local modules with `nf-core/modules`
- [x] updated `nanoseq` `dsl2` structure using `nf-core/rna-seq` as a reference
    - Here is [the spreadsheet of the local modules and subworkflows with links to their `versions.yml` update](https://docs.google.com/spreadsheets/d/1pPQkjkZ0d0Y5tM7j6NEV1kPCWl5GJM1p9Jg1km-jqNc/edit?usp=sharing)
- [x] tested the updated `nanoseq` with test data which [produced consistent results with the dsl1 pipeline](https://github.com/nf-core/nanoseq/files/7441320/compare_dsl1_newdsl2.tar.gz)
- [x] test the updated `nanoseq` with full data (`test_full`). The results are on `aws` machine, and downloading them will cost quite a bit, so it will be great if the result files can be transferred to the `awsmetadata` s3 bucket.
![Screenshot from 2021-10-29 16-13-35](https://user-images.githubusercontent.com/41866052/139422639-361d4fd2-de09-4be4-94a5-7d0a6e4a88cf.png)

christopher-hakkaart commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,459 @@
+ */
+include { QCBASECALL_PYCOQC_NANOPLOT       } from '../subworkflows/nf-core/qcbasecall_pycoqc_nanoplot'      addParams( pycoqc_options: pycoqc_options, nanoplot_options: nanoplot_options )
+include { QCFASTQ_NANOPLOT_FASTQC          } from '../subworkflows/nf-core/qcfastq_nanoplot_fastqc'         addParams( nanoplot_options: nanoplot_options, fastqc_options: fastqc_options )
+////////////////////////////////////////////////////
```
Add gap here

yuukiiwa replied:
I have added a gap

christopher-hakkaart commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,459 @@
+     //              Please specify the '--skip_quantification' parameter if you wish to skip these steps."""
+     //  }
+     //  REPLICATES_EXIST    = ch_sample.map { it -> it[0].split('_')[-1].replaceAll('R','').toInteger() }.max().val > 1
+     //  MULTIPLE_CONDITIONS = ch_sample.map { it -> it[0].split('_')[0..-2].join('_') }.unique().count().val > 1
```
Are these comments still needed or can they be deleted?

yuukiiwa replied:
I think I will keep this comment for a little longer after the transcript quantification part of the pipeline is updated in a later release

christopher-hakkaart reviewed (approved):
It looks really good. A few cosmetic notes but I didn't find anything else that would need fixed before a merge. Really great work.

yuukiiwa commented on modules/local/guppy.nf:
```diff
@@ -0,0 +1,71 @@
+        $proc_options \\
+        $barcode_ends \\
+        $config \\
+       $model
```
I have added a gap

yuukiiwa commented on modules/local/check_samplesheet.nf:
```diff
@@ -0,0 +1,28 @@
+    output:
+    path '*.csv'
+
+    script:  // This script is bundled with the pipeline, in nf-core/rnaseq/bin/
```
I have changed this to `nf-core/nanoseq/bin`

yuukiiwa commented on modules/local/minimap2_index.nf:
```diff
@@ -0,0 +1,45 @@
+        -t $task.cpus \\
+        -d ${fasta}.mmi \\
+        $fasta
+    ps
```
I have removed `ps`

yuukiiwa commented:
Thank you so much for the review, @christopher-hakkaart! I have addressed your comment in this PR: https://github.com/nf-core/nanoseq/pull/124

yuukiiwa commented:
Hi @christopher-hakkaart, I have also updated the CHANGELOG.md, which is currently on this PR: https://github.com/nf-core/nanoseq/pull/125
Thank you so much for reviewing!

cying111 commented on .github/workflows/branch.yml:
```diff
@@ -23,13 +23,22 @@ jobs:
             Hi @${{ github.event.pull_request.user.login }},
 
-            It looks like this pull-request is has been made against the ${{github.event.pull_request.head.repo.full_name}} `master` branch.
+            It looks like this pull-request is has been made against the [${{github.event.pull_request.head.repo.full_name }}](https://github.com/${{github.event.pull_request.head.repo.full_name }}) `master` branch.
```
is being made or has been made?
```suggestion
            It looks like this pull-request has been made against the [${{github.event.pull_request.head.repo.full_name }}](https://github.com/${{github.event.pull_request.head.repo.full_name }}) `master` branch.
```

yuukiiwa replied:
I removed "is" from the sentence

cying111 commented on CHANGELOG.md:
```diff
@@ -3,6 +3,33 @@
+|-------------------------|-------------|-------------|
+| `bioconductor-bambu`    | 1.0.0       | 1.0.2       |
+| `nanolyse`              |             | 1.2.0       |
+| `r-base`                | 4.0.3       | 4.0.2       |
```
Just curious: why is there is change back in r-base version? Anything that didn't run in 4.0.3?

yuukiiwa replied:
The [r-base version was changed](https://github.com/nf-core/nanoseq/commit/b5bf1bc861602163fd4f02fcaabac045d036654c#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed) because `DEXseq` in r-4.0.3 did not work with the test data

cying111 commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,46 +1,111 @@
+## Preamble
 
-Examples of unacceptable behavior by participants include:
+> Note: This Code of Conduct (CoC) has been drafted by the nf-core Safety Officer and been edited after input from members of the nf-core team and others. "We", in this document, refers to the Safety Officer and members of the nf-core core team, both of whom are deemed to be members of the nf-core community and are therefore required to abide by this Code of Conduct. This document will amended pe ...
```
```suggestion
> Note: This Code of Conduct (CoC) has been drafted by the nf-core Safety Officer and been edited after input from members of the nf-core team and others. "We", in this document, refers to the Safety Officer and members of the nf-core core team, both of whom are deemed to be members of the nf-core community and are therefore required to abide by this Code of Conduct. This document will be amended periodically to keep it up-to-date, and in case of any dispute, the most current version will apply.
```

yuukiiwa replied:
I have added "be" to the sentence

cying111 commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,46 +1,111 @@
+
+We have therefore adopted this code of conduct (CoC), which we require all members of our community and attendees in nf-core events to adhere to in all our workspaces at all times. Workspaces include but are not limited to Slack, meetings on Zoom, Jitsi, YouTube live etc.
+
+Our CoC will be strictly enforced and the nf-core team reserve the right to exclude participants who do not comply with our guidelines from our workspaces and future nf-core activities.
```
```suggestion
Our CoC will be strictly enforced and the nf-core team reserves the right to exclude participants who do not comply with our guidelines from our workspaces and future nf-core activities.
```

yuukiiwa replied:
I have changed "reserve" to "reserves"

cying111 commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,46 +1,111 @@
+
+Members of the core team or the safety officer who violate the CoC will be required to recuse themselves pending investigation. They will not have access to any reports of the violations and be subject to the same actions as others in violation of the CoC.
+
+## When are where does this Code of Conduct apply?
```
```suggestion
## When and where does this Code of Conduct apply?
```

yuukiiwa replied:
I have changed "are" to "and"

cying111 reviewed (changes requested):
I only had very minor comments regarding some of the texts and one curiosity question. Other than that all look good to me! Great work @yuukiiwa !

yuukiiwa commented:
Hi @cying111, thank you so much for reviewing the PR! I have addressed your suggestions in this PR (https://github.com/nf-core/nanoseq/pull/126), but because of the changes made, I got this error in the linting checks, so I guess we will have to keep it as it is and ask `nf-core` to change those typos:
![Screenshot 2021-11-22 at 11 59 51 AM](https://user-images.githubusercontent.com/41866052/142799158-4d234f26-9de6-435f-b255-4ca35a5f3fff.png)

drpatelh commented on nextflow_schema.json:
```diff
@@ -435,28 +241,126 @@
         },
-        {
-            "$ref": "#/definitions/institutional_config_options"
+        "barcode_kit": {
+            "type": "string"
+        },
+        "barcode_both_ends": {
+            "type": "string"
+        },
+        "guppy_config": {
+            "type": "string"
+        },
+        "guppy_model": {
+            "type": "string"
+        },
+        "guppy_gpu": {
+            "type": "string"
+        },
+        "guppy_gpu_runners": {
+            "type": "integer",
+            "default": 6
```
These all need to be added properly to the Schema via `nf-core schema build`. Looks like I had already added descriptions, logos already but they were deleted?

yuukiiwa commented:
aws test:
![Screenshot 2021-11-25 at 9 53 31 PM](https://user-images.githubusercontent.com/41866052/143453830-a7bf0dce-aada-411e-b967-1d096f5e78f5.png)

jfy133 commented on docs/output.md:
```diff
@@ -1,9 +1,5 @@
 # nf-core/nanoseq: Output
 
-## :warning: Please read this documentation on the nf-core website: [https://nf-co.re/nanoseq/output](https://nf-co.re/nanoseq/output)
```
Is this removal from the template? :thinking:

yuukiiwa replied:
I have added this back in

jfy133 commented on docs/output.md:
```diff
@@ -21,51 +17,67 @@ The directories listed below will be created in the output directory after the p
+[NanoLyse](https://github.com/wdecoster/nanolyse)
+
+*Description*:
+If you would like to run NanoLyse on the raw FastQ files then you can provide `--run_nanolyse` when running the pipeline. By default, the pipeline will filter the raw reads relative to lambda phage but you can provide your own fasta file of "contaminants" with `--nanolyse_fasta`. The filtered FastQ files will be used in the downstream processes of the pipeline.
```
This reads more as 'usage' information rather than output? Maybe you could rephrase to just say what the filtered FASTQ will contain if you run X vs Z.

yuukiiwa replied:
I have changed the description to "The filtered FastQ files will contain raw reads without the provided reference sequences (default: lambda phage sequences)."

jfy133 replied:
I was mostly referring to the earlier sentences, but this OK for now.

jfy133 commented on docs/usage.md:
```diff
@@ -156,11 +156,55 @@ nextflow run nf-core/nanoseq \
+| Column         | Description                                                                                                                                                                            |
+|----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `sample`       | Custom sample name. This entry will be identical for multiple sequencing libraries/runs from the same sample. Spaces in sample names are automatically converted to underscores (`_`). |
+| `fastq_1`      | Full path to FastQ file for Illumina short reads 1. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".                                                             |
+| `fastq_2`      | Full path to FastQ file for Illumina short reads 2. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".                                                             |
```
Shouldn't this be nanopore data rather than Illumina? Or is this all boilerplate from the template - should be updated if so.

yuukiiwa replied:
I have removed this table. The correct table is in **Introduction**

jfy133 replied:
It will be iportant to look through `usage.md` in the future and re-write for nanoseq as far as possible :+1:

jfy133 commented on modules/local/bam_rename.nf:
```diff
@@ -0,0 +1,17 @@
+    tuple val(meta), path(bam)
+
+    output:
+    tuple val(meta), path("*.bam"), emit: sortbam_quant
```
```suggestion
    tuple val(meta), path("*.bam"), emit: sortbam_quant
```

a more informative output channel might be good here (I guess copy paste error)

yuukiiwa replied:
I have changed this to bam

jfy133 commented on modules/local/bedtools_genomecov.nf:
```diff
@@ -0,0 +1,34 @@
+//params.options = [:]
+//def options    = initOptions(params.options)
+
+process BEDTOOLS_GENOMECOV {
```
Fine as local module for now, but there is now an 'official' nf-core/modules for [this](https://github.com/nf-core/modules/blob/master/modules/bedtools/genomecov/main.nf) to consider for the future (same ofr a few of the other modules here)

yuukiiwa replied:
I will use the bedtools genomecov modules from nf-core/modules in a later release

jfy133 commented on modules/local/get_chrom_sizes.nf:
```diff
@@ -0,0 +1,35 @@
+//        saveAs: { filename -> saveFiles(filename:filename, publish_dir:"genome") }
+
+    conda     (params.enable_conda ? "bioconda::samtools=1.10" : null)
+    container "quay.io/biocontainers/samtools:1.10--h9402c20_2"
```
Just noticed: does this quay link also support Singularity? Asking given the nf-core/modules always have two links - one from quay (docker) and the other from galaxy (singularity).

Sorry if this has been discussed before 😅

jfy133 replied:
I see other local modules have such options along the same lines (e.g. get_nanolyse_fasta)

yuukiiwa replied:
I have updated this to the following:
```
    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
        container "https://depot.galaxyproject.org/singularity/samtools:1.13--h8c37831_0"
    } else {
        container "quay.io/biocontainers/samtools:1.13--h8c37831_0"
    }
```

jfy133 commented on modules/local/get_test_data.nf:
```diff
@@ -0,0 +1,18 @@
+    script:
+    barcoded = workflow.profile.contains('test_bc_nodx') ? "nonbarcoded" : "barcoded"
+    """
+    git clone https://github.com/nf-core/test-datasets.git --branch nanoseq --single-branch
```
What is the purpose of this module? Why not give direct raw github links in the test data input samplesheet?

yuukiiwa replied:
This is for staging [all test fast5 files](https://github.com/nf-core/test-datasets/tree/nanoseq/fast5/barcoded) from `--input_path [fast5 directory]`

jfy133 replied:
Ah, input directory - I see. OK!

For nf-core/eager we just cloned the test-data repository as a GitHub Actions CI step, but I guess nice to bundle it with the pipeline itself for others to use :+1:

jfy133 commented on modules/local/graphmap2_index.nf:
```diff
@@ -0,0 +1,41 @@
+
+    script:
+    def preset = (params.protocol == 'DNA' || is_transcripts) ? "" : "-x rnaseq"
+    // TODO pipeline: Should be staging gtf file properly as an input
```
```suggestion
```

yuukiiwa replied:
I have removed this line

jfy133 commented on modules/local/guppy.nf:
```diff
@@ -0,0 +1,71 @@
+    ## Concatenate fastq files
+    mkdir fastq
+    cd basecalling
+    if [ "\$(find . -type d -name "barcode*" )" != "" ]
```
May want to be careful with single sqaure bracket quotes in bash if statements, I've had issues where this would break on some clusters

yuukiiwa replied:
I will look into this in a later release because this may take quite a bit of time to modify and test

jfy133 replied:
OK, fine with me. Just something to keep in mind in the future if someone hits an error on that particular module!

jfy133 commented on modules/local/minimap2_align.nf:
```diff
@@ -0,0 +1,45 @@
+    def preset    = (params.protocol == 'DNA' || is_transcripts) ? "-ax map-ont" : "-ax splice"
+    def kmer      = (params.protocol == 'directRNA') ? "-k14" : ""
+    def stranded  = (params.stranded || params.protocol == 'directRNA') ? "-uf" : ""
+    // TODO pipeline: Should be staging bed file properly as an input
```
```suggestion
```

yuukiiwa replied:
I have removed this line

jfy133 reviewed (approved):
Technically this looks sound for a first release, however:

The usage doucmentation (`docs/usage.md`) also appears to be all the boilerplate from the template (i.e. RNAseq). I would recommend at least updating all the samplesheet sections, as presumably this would not apply to your pipeline.

A lot of the output documentation  () reads as usage/input specification though (i.e., saying use this parameter for this purpose, and this parameter for this purpose etc.), so I would recommend in the future considering to rewrite the documentation.

Another main question is why a lot of the local modules only have docker images defined (AFAICS) and not singluarity URLs from Galaxy? Does the pipeline work for at least all three of `-profile conda/docker/singulairty`?

There are quite a few TODOs regarding staging bed files, they either should be removed or resolved if they are not already

jfy133 commented on modules/local/samplesheet_check.nf:
```diff
@@ -0,0 +1,31 @@
+// Import generic module functions
+include { saveFiles } from './functions'
```
Difference between this and `check_samplesheet.nf`?

yuukiiwa replied:
There is not much difference. I will keep `samplesheet_check.nf` because it is from the template update

jfy133 replied:
:+1:

jfy133 commented on modules/local/ucsc_bed12tobigbed.nf:
```diff
@@ -0,0 +1,42 @@
+
+    conda     (params.enable_conda ? "bioconda::ucsc-bedtobigbed=377" : null)
+    container "quay.io/biocontainers/ucsc-bedtobigbed:377--h446ed27_1"
+    if ({ task.exitStatus in [255] }) { errorStrategy = 'ignore' }
```
Would this not be better in `base.config`?

yuukiiwa replied:
The reason this is here is that base.config has already defined  `errorStrategy = { task.exitStatus in [143,137,104,134,139] ? 'retry' : 'finish' }`

jfy133 replied:
OK, but you can add additional custom exit codes in base.config on a per process basis:

DSL1 example: https://github.com/nf-core/eager/blob/cc66639a8a2710f0de6aef5ae28e1614a5a4e3e2/conf/base.config#L93-L95

I suggest to do this to keep it all in one place, rather than spread all over the place.

jfy133 commented on nextflow.config:
```diff
@@ -1,202 +1,252 @@
-  nextflowVersion = '>=19.10.0'
-  version = '1.1.0'
+    name            = 'nf-core/nanoseq'
+    author          = 'Chelsea Sawyer'
```
```suggestion
    author          = 'Chelsea Sawyer, Yuk Kei Wan'
```

yuukiiwa replied:
I have added myself in

jfy133 commented on subworkflows/local/quantify_stringtie_featurecounts.nf:
```diff
@@ -0,0 +1,64 @@
+        .set { ch_sample_gtf }
+
+    /*
+     * Merge isoforms across samples caleed by StringTie
```
```suggestion
     * Merge isoforms across samples called by StringTie
```

yuukiiwa replied:
I have fixed this typo

jfy133 commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,456 @@
+multiqc_options.args       += params.multiqc_title ? " --title \"$params.multiqc_title\"" : ''
+if (params.skip_alignment)  { multiqc_options['publish_dir'] = '' }
+
+// TO DO -- define options for the processes below
```
```suggestion
```

yuukiiwa replied:
I have removed this line

jfy133 commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,456 @@
+/*
+ * SUBWORKFLOW: Consisting of a mix of local and nf-core/modules
+ */
+// TO DO -- define options for the subworkflows below
```
```suggestion
```

yuukiiwa replied:
I have removed this line

jfy133 commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,456 @@
+/*
+ * MODULE: Installed directly from nf-core/modules
+ */
+// TO DO -- define options for the processes below
```
```suggestion
```

yuukiiwa replied:
I have removed this line

jfy133 commented on workflows/nanoseq.nf:
```diff
@@ -0,0 +1,456 @@
+        MULTIPLE_CONDITIONS = false
+        ch_sample.map{ it[2] }.unique().toList().set { fastas }
+        ch_sample.map{ it[3] }.unique().toList().set { gtfs }
+        // BUG: ".val" halts the pipeline ///////////////////////
```
Does this error still apply?

yuukiiwa replied:
These no longer apply. I'm keeping them there as a note.

jfy133 replied:
OK :+1:

jfy133 reviewed (commented):
A few more comments, see the first review above

yuukiiwa commented on CHANGELOG.md:
```diff
@@ -3,6 +3,33 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [2.0.0] - 2021-11-22
```
I think I will keep it as it is here just to be consistent with the previous releases. I do have a name in mind -- Green Sloth

yuukiiwa commented on CHANGELOG.md:
```diff
@@ -3,6 +3,33 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [2.0.0] - 2021-11-22
```
I have applied the changes

yuukiiwa commented on README.md:
```diff
@@ -18,61 +18,71 @@
 
 **nfcore/nanoseq** is a bioinformatics analysis pipeline that can be used to perform basecalling, demultiplexing, QC, mapping and downstream analysis of Nanopore DNA/RNA sequencing data.
```
I have applied the changes

yuukiiwa commented:
Thank you so much for the suggestions, @jfy133! 
I don't have write access to the `dev` branch so I could commit your suggestions directly, so I am addressing them from PR https://github.com/nf-core/nanoseq/pull/132

jfy133 commented on docs/usage.md:
```diff
@@ -192,12 +192,6 @@ TREATMENT_REP3,AEG588A6_S6_L003_R1_001.fastq.gz,
-| `fastq_1`      | Full path to FastQ file for Illumina short reads 1. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".                                                             |
-| `fastq_2`      | Full path to FastQ file for Illumina short reads 2. File has to be gzipped and have the extension ".fastq.gz" or ".fq.gz".                                                             |
-
 An [example samplesheet](../assets/samplesheet.csv) has been provided with the pipeline.
```
Does this example also contain a `nanoseq` example?

yuukiiwa replied:
Yes, [`../assets/samplesheet.csv`](https://github.com/nf-core/nanoseq/blob/dev/assets/samplesheet.csv) is an example samplesheet.csv specifically for `nanoseq`

jfy133 commented on modules/local/get_chrom_sizes.nf:
```diff
@@ -12,7 +12,11 @@ process GET_CHROM_SIZES {
 
     conda     (params.enable_conda ? "bioconda::samtools=1.10" : null)
-    container "quay.io/biocontainers/samtools:1.10--h9402c20_2"
+    if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
+        container "https://depot.galaxyproject.org/singularity/samtools:1.13--h8c37831_0"
+    } else {
+        container "quay.io/biocontainers/samtools:1.13--h8c37831_0"
+    }
```
Please check this syntax is used for all internal modules

yuukiiwa replied:
I have updated all local modules that use quay containers with the syntax above. Customized containers were made for the processes for `BAMBU` (will be updated in the next release), `DEXSeq`(also contain other packages in the process), and `GET_TEST_DATA` (couldn't find a container with `git`) are not in this syntax

jfy133 reviewed (commented):
I noticed a couple of other things. Main thing would be checking all local modules using container have the same syntax

yuukiiwa commented:
Thanks for the suggestions, @jfy133! I have updated the container syntax for the processes that use a `quay` container and moved the `errorStrategy` for the UCSC tools to `base.config`. The changes were made in PR https://github.com/nf-core/nanoseq/pull/133
